### DISCUSSION TITLE
Wrap an api around a third-party crate

### DISCUSSION THREAD

**Discussion by Roms1383** [fzyzcjy/flutter_rust_bridge#275] [Q&A] [answered]
Best wishes for 2022 🎉 

I wanted to add ambient sounds on my hobby project, so I got a quick PoC using [kira](https://crates.io/crates/kira) which allows to play, mix and transform sounds. I chose `kira` mainly because it's great, but also because I couldn't find any comparable alternative on pub.dev.

Long story short, I tried a bit but couldn't get anything to compile and I'm wondering if using `flutter_rust_bridge` to manipulate a [kira::manager::AudioManager](https://docs.rs/kira/0.5.3/kira/manager/struct.AudioManager.html) over FFI is possible, or if this an inappropriate use case for `flutter_rust_bridge` ?

**Comment by fzyzcjy** [comment, marked as answer]
Interesting! I guess yes. We already support rich features (have a look at frb_example/pure_dart/rust/api.rs to see tests) and will support more. Even if you find something that does not exist yet, you can always make a PR and I almost always respond quickly!

**Reply by Roms1383** [reply]
Thanks!
Well TBH I already did, but got stuck.
I'm somehow familiar with Rust but not with FFI and Flutter and got stuck.
My initial idea was to keep the `AudioManager` around, behind a `Mutex` with `lazy_static` and to provide, for a quick test, with something like:

```rust
// api.rs

lazy_static! {
    pub static ref AUDIO: Mutex<AudioManager<CpalBackend>> = {
        Mutex::new(
            AudioManager::new(
                CpalBackend::new().expect("audio cpal backend"),
                AudioManagerSettings::default(),
            )
            .expect("create audio manager"),
        )
    };
}

#[repr(u8)]
pub enum Screen {
    Splash,
    Home,
}

#[no_mangle]
pub extern "C" fn play(screen: Screen) {
    let filename = match screen {
        Screen::Home => "../assets/sound/home-screen.mp3",
        Screen::Splash => "../assets/sound/splash-screen.mp3",
    };
    let sound = kira_loaders::load(filename, StaticSoundSettings::default());
    if let Ok(sound) = sound {
        match AUDIO.lock().expect("static audio manager").play(sound) {
            Ok(_) => {}
            Err(_) => {}
        }
    }
}
```
Also finding a way to return some output so that Flutter app knows if things go right or not.

**Reply by fzyzcjy** [reply]
`#[no_mangle] pub extern "C"` no need for that. have a look at tutorial and frb_example/with_flutter and frb_example/pure_rust

**Reply by fzyzcjy** [reply]
that proto sounds good. I may even put more things in Flutter (to enjoy its fast hot reload): e.g. `play(filename: String, other_arguments: Sth)`, but of course it is your choice

**Reply by Roms1383** [reply]
yes since this is just my hobby project I experiment but can't spend as much time as I would like to,
so I try to get small things working fast first and iterate afterwards if successful. I'll try to give it another shot after getting back to the examples again, thanks.

**Reply by Roms1383** [reply]
I ended up making a quick [sample repo](https://github.com/Roms1383/test-flutter-rust-bridge) and opening an [issue describing the error](https://github.com/Roms1383/test-flutter-rust-bridge/issues/1) I encountered.
